src/executor/executor.c

```diff
@@ -5239,14 +5266,65 @@ int main(int argc, char *argv[]) {
                 log_open();
         }
 
-        r = fdset_remove(fdset, fileno(arg_serialization));
-        if (r < 0)
-                return log_error_errno(r, "Failed to remove serialization fd from fd set: %m");
+        /* If we are forked directly then we will have all the required FDs already open, create a set to
+         * prepare for deserialization. If we are getting a socket instead we'll get them in an array via
+         * SCM_RIGHTS. */
+        if (arg_serialization) {
+                serialization = arg_serialization;
+
+                r = fdset_new_fill(/* filter_cloexec= */ 0, &fdset);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create fd set: %m");
+        } else if (arg_manager_socket >= 0) {
+                _cleanup_close_ int serialization_fd = -EBADF, pidfd = -EBADF;
 
-        r = exec_deserialize_invocation(arg_serialization,
+                manager_socket = arg_manager_socket;
+
+                /* Wait for serialization FD from manager datagram socket */
+                r = receive_many_fds(manager_socket, &fds_array, &n_fds_array, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to receive serialization FDs: %m");
+
+                /* There always at the very least the exec FD */
+                if (n_fds_array == 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Received no FDs from manager socket.");
+
+                /* The memfd with all the data is always the first one */
+                serialization_fd = TAKE_FD(fds_array[0]);
+
+                log_debug("Accepting work from worker " PID_FMT, getpid_cached());
+
+                /* Send back pidfd so that systemd knows who took up this job */
+                pidfd = pidfd_open(getpid_cached(), 0);
+                if (pidfd < 0) {
+                        char iov_buffer[DECIMAL_STR_MAX(pid_t) + 1] = {};
+                        struct iovec iov = IOVEC_MAKE(iov_buffer, sizeof(iov_buffer) - 1);
+
+                        if (!ERRNO_IS_NOT_SUPPORTED(errno))
+                                return log_error_errno(errno, "Failed to open pidfd: %m");
+
+                        /* Fallback to sending the pid */
+                        xsprintf(iov_buffer, "%d", getpid_cached());
+                        r = send_one_fd_iov(manager_socket, /* fd= */ -EBADF, &iov, 1, 0);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to send pid: %m");
+                } else {
+                        r = send_one_fd(manager_socket, pidfd, 0);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to send pidfd: %m");
+                }
+
```

Comment:
idea, instead of sending the pidfd up, just send an error `int` up, where 0 means success. And then read the pid from the datagram metadata. 

i think it's generally a better idea to return an error code here back to the parent, maybe one day the executor wants to return an error already when it figured out it can't work, even if it doesn't do that for now.

Reply:
That would mean using pidfd to track the process only in the very latest kernel. The current implementation instead allows to do that on the latest 3 LTS kernels too. If one day sending an error code is needed we can revisit, and then the SO_PEERPIDFD will be available more widely, but for now I don't think it's worth losing that given it's not needed